File: scala/revenj-core/src/main/scala/net/revenj/extensibility/PluginLoader.scala
package net.revenj.extensibility

import net.revenj.Utils

import scala.reflect.runtime.universe._

trait PluginLoader {
  def find[T : TypeTag]: scala.collection.Seq[Class[T]]

  @deprecated("Use resolve without class argument", "0.5.0")
  def resolve[T : TypeTag](container: Container, manifest: Class[T]): Array[T] = {
    resolve[T](container)
  }
  def resolve[T : TypeTag](container: Container): Array[T] = {
    val loader = container.resolve[ClassLoader]
    val mirror = runtimeMirror(loader)
    Utils.findType(mirror.typeOf[T], mirror) match {
      case Some(tpe) =>
        val scope = container.createScope()
        try {
          find[T] foreach { sc =>
            scope.registerType(tpe, sc, InstanceScope.Context)
          }
          scope.resolve[Array[T]]
        } finally {
          scope.close()
        }
      case _ => throw new IllegalArgumentException(s"Unable to resolve plugins for provided type: ${mirror.typeOf[T]}")
    }
  }
}

File: java/revenj-core/src/main/java/org/revenj/extensibility/PluginLoader.java
package org.revenj.extensibility;

import java.lang.reflect.GenericArrayType;
import java.lang.reflect.Type;
import java.util.List;

public interface PluginLoader {
	<T> List<Class<T>> find(Class<T> manifest, Type... types) throws Exception;

	@SuppressWarnings("unchecked")
	default <T> T[] resolve(Container container, Class<T> manifest) throws Exception {
		try (Container scope = container.createScope()) {
			List<Class<T>> manifests = find(manifest);
			for (Class<T> sc : manifests) {
				scope.registerType(manifest, sc, InstanceScope.CONTEXT);
			}
			return (T[]) scope.resolve((GenericArrayType) () -> manifest);
		}
	}
}
